src/Microsoft.AspNetCore.Blazor/Components/Provider.cs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Blazor.RenderTree;
+using System;
+
+namespace Microsoft.AspNetCore.Blazor.Components
+{
+    /// <summary>
+    /// A component that provides one or more parameter values to all descendant components.
+    /// </summary>
+    public class Provider : IComponent
+    {
+        private RenderHandle _renderHandle;
+
+        /// <summary>
+        /// The content to which the value should be provided.
+        /// </summary>
+        [Parameter] private RenderFragment ChildContent { get; set; }
+
+        /// <summary>
+        /// The value to be provided.
+        /// </summary>
+        [Parameter] private object Value { get; set; }
+
+        /// <summary>
+        /// Optionally gives a name to the provided value. Descendant components
+        /// will be able to receive the value by specifying this name.
+        ///
+        /// If no name is specified, then descendant components will receive the
+        /// value based the type of value they are requesting.
+        /// </summary>
+        [Parameter] private string Name { get; set; }
+
+        /// <inheritdoc />
+        public void Init(RenderHandle renderHandle)
+        {
+            _renderHandle = renderHandle;
+        }
+
+        /// <inheritdoc />
+        public void SetParameters(ParameterCollection parameters)
+        {
+            // Implementing the parameter binding manually, instead of just calling
+            // parameters.AssignToProperties(this), is just a very slight perf optimization
+            // and makes it simpler impose rules about the params being required or not.
+
+            var hasSuppliedValue = false;
+            Value = null;
+            ChildContent = null;
+            Name = null;
+
+            foreach (var parameter in parameters)
+            {
+                if (parameter.Name.Equals(nameof(Value), StringComparison.OrdinalIgnoreCase))
+                {
+                    Value = parameter.Value;
+                    hasSuppliedValue = true;
+                }
+                else if (parameter.Name.Equals(nameof(ChildContent), StringComparison.OrdinalIgnoreCase))
+                {
+                    ChildContent = (RenderFragment)parameter.Value;
+                }
+                else if (parameter.Name.Equals(nameof(Name), StringComparison.OrdinalIgnoreCase))
+                {
+                    Name = (string)parameter.Value;
+                    if (string.IsNullOrEmpty(Name))
+                    {
+                        throw new ArgumentException($"The parameter '{nameof(Name)}' for component '{nameof(Provider)}' does not allow null or empty values.");
```

Comment:
nit: there's an argument (pun intended) to be made for using another exception type here since theres error aren't strictly related to the arguments of this method. I tend not to use `ArgumentException` unless I have an argument name to provide.

I don't think it's important, but rather we should make a decision and then be consistent for the built-in components. I suspect it is already consistent since you wrote them!

Reply:
I have been using `ArgumentException` in the broader sense of considering component arguments to be arguments, even though they aren't CLR method args. Totally accept the point that if someone thinks `ArgumentException` precisely means CLR method arguments then this would be confusing.

If we want to change this later I'd be fine with that. Since you've said you don't think it's important, and I'm OK with it, I'll leave it as-is.

Reply:
Yeah it's something to think about. Deciding that 'error-related-to-component-parameter' is a first class concept might be valuable in the future for diagnostics and logging.